File: deploy_1val.mjs
/**
 * Deploy sports_lib.py to Bradbury using 1 validator (instead of default 5).
 * This gives a much better chance of succeeding if at least 1 validator is online.
 */
import { readFileSync } from 'fs';
import { createWalletClient, createPublicClient, http, parseGwei } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { ethers } from 'ethers';

const { Wallet, JsonRpcProvider, Interface, toUtf8Bytes, keccak256 } = ethers;

const RPC = 'https://rpc-bradbury.genlayer.com';
const CONSENSUS = '0x0112Bf6e83497965A5fdD6Dad1E447a6E004271D';
const CHAIN_ID = 4221;

// Load keystore
const ks = JSON.parse(readFileSync('./tmp_key.json', 'utf8'));
const provider = new JsonRpcProvider(RPC, CHAIN_ID, { staticNetwork: true });
const wallet = await Wallet.fromEncryptedJson(JSON.stringify(ks), 'password123');
const signer = wallet.connect(provider);

console.log('Deployer:', await signer.getAddress());

// Read contract code
const code = readFileSync('./sports_lib.py');
const codeHex = '0x' + code.toString('hex');

// GL encode helpers
function writeULEB(out, n) {
  n = BigInt(n);
  while (true) {
    let b = Number(n & 0x7fn);
    n >>= 7n;
    if (n) { out.push(b | 0x80); } else { out.push(b); break; }
  }
}
function encodeVal(out, v) {
  if (v === null) { writeULEB(out, 0n); }
  else if (typeof v === 'string') {
    const b = new TextEncoder().encode(v);
    writeULEB(out, BigInt(b.length << 3) | 4n);
    out.push(...b);
  } else if (Array.isArray(v)) {
    writeULEB(out, BigInt(v.length << 3) | 5n);
    v.forEach(x => encodeVal(out, x));
  } else if (typeof v === 'object') {
    const entries = Object.entries(v).sort(([a], [b]) => a < b ? -1 : 1);
    writeULEB(out, BigInt(entries.length << 3) | 6n);
    for (const [k, val] of entries) {
      const kb = new TextEncoder().encode(k);
      writeULEB(out, BigInt(kb.length));
      out.push(...kb);
      encodeVal(out, val);
    }
  }
}
function rlpBytes(b) {
  b = Uint8Array.from(b);
  if (b.length === 1 && b[0] < 0x80) return b;
  if (b.length < 56) return Uint8Array.from([0x80 + b.length, ...b]);
  const lb = [];
  let l = b.length;
  while (l > 0) { lb.unshift(l & 0xff); l >>= 8; }
  return Uint8Array.from([0xb7 + lb.length, ...lb, ...b]);
}
function rlpList(items) {
  const payload = items.flatMap(x => [...rlpBytes(x)]);
  if (payload.length < 56) return Uint8Array.from([0xc0 + payload.length, ...payload]);
  const lb = [];
  let l = payload.length;
  while (l > 0) { lb.unshift(l & 0xff); l >>= 8; }
  return Uint8Array.from([0xf7 + lb.length, ...lb, ...payload]);
}
function glSerialize(obj) {
  const inner = [];
  encodeVal(inner, obj);
  const rlp = rlpList([Uint8Array.from(inner), Uint8Array.from([0x00])]);
  return '0x' + Array.from(rlp, b => b.toString(16).padStart(2, '0')).join('');
}

// Deploy calldata: constructor call
const deployCalldata = glSerialize({ method: '__init__', args: [] });

// addTransaction ABI — 6 args
const ABI = new Interface([
  'function addTransaction(address _sender, address _recipient, uint256 _numOfInitialValidators, uint256 _maxRotations, bytes _txData, uint256 _validUntil)'
]);

// Encode the inner deploy payload: [codeHex, args=[], leaderOnly=false]
// The txData for a deploy is: rlp([code_bytes, calldata_bytes, leaderOnly_bool])
function rlpBytesFromHex(hex) {
  const b = Uint8Array.from(Buffer.from(hex.slice(2), 'hex'));
  return rlpBytes(b);
}
function rlpBool(v) {
  return rlpBytes(Uint8Array.from([v ? 0x01 : 0x00]));
}

// txData = RLP([code, deployCalldata, leaderOnly])
const codeBytes = Buffer.from(codeHex.slice(2), 'hex');
const calldataBytes = Buffer.from(deployCalldata.slice(2), 'hex');
const innerRlp = rlpList([codeBytes, calldataBytes, new Uint8Array([0x00])]);
const txData = '0x' + Buffer.from(innerRlp).toString('hex');

const senderAddr = await signer.getAddress();
const deployAddr = '0x0000000000000000000000000000000000000000'; // null address = deploy

const data = ABI.encodeFunctionData('addTransaction', [
  senderAddr,
  deployAddr,
  1n,   // 1 validator only!
  0n,   // 0 max rotations
  txData,
  0n    // validUntil = 0 (no expiry)
]);

console.log('Sending deploy with 1 validator...');
const nonce = await provider.getTransactionCount(senderAddr);
console.log('Nonce:', nonce);

const tx = await signer.sendTransaction({
  to: CONSENSUS,
  data,
  gasLimit: 2000000n,
  gasPrice: parseGwei('1'),
  nonce,
  chainId: CHAIN_ID,
  type: 0
});

console.log('ETH TX hash:', tx.hash);
console.log('Waiting for ETH receipt...');
const receipt = await tx.wait();
console.log('ETH status:', receipt.status);
console.log('Logs:', receipt.logs.length, 'events');
if (receipt.logs.length > 0) {
  console.log('First log topics:', receipt.logs[0].topics);
}

File: deploy_1val.py
"""
Deploy sports_lib.py to Bradbury with 1 validator.
Uses eth_account to sign, web3 to send.
"""
import json, sys
from eth_account import Account
from web3 import Web3
from eth_abi import encode as abi_encode

RPC = 'https://rpc-bradbury.genlayer.com'
CONSENSUS = '0x0112Bf6e83497965A5fdD6Dad1E447a6E004271D'
CHAIN_ID = 4221

# Load keystore
with open('tmp_key.json') as f:
    ks = json.load(f)
account = Account.from_key(Account.decrypt(ks, 'password123'))
print(f'Deployer: {account.address}')

w3 = Web3(Web3.HTTPProvider(RPC, request_kwargs={'headers': {'Origin':'http://localhost:3000','User-Agent':'Mozilla/5.0'}}))
print(f'Connected: {w3.is_connected()}')

# Read contract code
with open('sports_lib.py', 'rb') as f:
    code_bytes = f.read()
print(f'Code size: {len(code_bytes)} bytes')

# GL Encode helpers
def write_uleb(out, n):
    n = int(n)
    while True:
        b = n & 0x7F; n >>= 7
        if n: out.append(b | 0x80)
        else: out.append(b); return

def encode_val(out, v):
    if v is None:
        write_uleb(out, 0)
    elif isinstance(v, str):
        b = v.encode('utf-8')
        write_uleb(out, (len(b) << 3) | 4)
        out.extend(b)
    elif isinstance(v, list):
        write_uleb(out, (len(v) << 3) | 5)
        for x in v: encode_val(out, x)
    elif isinstance(v, dict):
        items = sorted(v.items())
        write_uleb(out, (len(items) << 3) | 6)
        for k, val in items:
            kb = k.encode('utf-8')
            write_uleb(out, len(kb))
            out.extend(kb)
            encode_val(out, val)

def rlp_enc_bytes(b):
    b = bytes(b)
    if len(b) == 1 and b[0] < 0x80: return b
    if len(b) < 56: return bytes([0x80 + len(b)]) + b
    l = len(b); lb = []
    while l: lb.insert(0, l & 0xff); l >>= 8
    return bytes([0xb7 + len(lb)] + lb) + b

def rlp_enc_list(items):
    payload = b''.join(rlp_enc_bytes(x) for x in items)
    if len(payload) < 56: return bytes([0xc0 + len(payload)]) + payload
    l = len(payload); lb = []
    while l: lb.insert(0, l & 0xff); l >>= 8
    return bytes([0xf7 + len(lb)] + lb) + payload

# Constructor calldata = encode4({}) = ULEB128(0 << 3 | 6) = 0x06 (empty map)
# matches genlayer-js: encode4(makeCalldataObject(undefined, [], {})) = encode4({})
constructor_calldata = bytes([0x06])

# txData = RLP([code_bytes, constructor_calldata, false=0x00])
# matches: serialize([code, encode4({}), false]) = toRlp([toHex(code), "0x06", "0x00"])
tx_data_inner = rlp_enc_list([code_bytes, constructor_calldata, b'\x00'])

# addTransaction ABI
ADD_TX_SIG = b'\xf4\x2a\x47\x72'  # keccak256("addTransaction(address,address,uint256,uint256,bytes,uint256)")[:4]
# Let me compute it:
import hashlib
sig_full = 'addTransaction(address,address,uint256,uint256,bytes,uint256)'
selector = Web3.keccak(text=sig_full)[:4]
print(f'Selector: {selector.hex()}')

# Encode parameters
sender = account.address
recipient = '0x0000000000000000000000000000000000000000'  # null = deploy
num_validators = 1
max_rotations = 0
valid_until = 0

# ABI encode the function call
from eth_abi import encode
params_encoded = encode(
    ['address', 'address', 'uint256', 'uint256', 'bytes', 'uint256'],
    [sender, recipient, num_validators, max_rotations, tx_data_inner, valid_until]
)
data = selector + params_encoded

# Get nonce
nonce = w3.eth.get_transaction_count(account.address)
print(f'Nonce: {nonce}')

# Build and sign tx
tx = {
    'to': Web3.to_checksum_address(CONSENSUS),
    'data': data,
    'gas': 2_000_000,
    'gasPrice': Web3.to_wei('1', 'gwei'),
    'nonce': nonce,
    'chainId': CHAIN_ID,
}

signed = account.sign_transaction(tx)
print('Sending TX...')
tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
print(f'TX hash: {tx_hash.hex()}')

print('Waiting for receipt...')
receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=60)
print(f'Status: {receipt.status}')
print(f'Block: {receipt.blockNumber}')
print(f'Logs: {len(receipt.logs)}')
for log in receipt.logs:
    print(f'  Log topic[0]: {log.topics[0].hex() if log.topics else "none"}')
    print(f'  Log data: {log.data.hex()[:100]}')
